### ISSUE TITLE
Stale processes remain when training a model with data preparation done using multiprocessing

### ISSUE THREAD

**Issue by vsenko** [fchollet/keras#8622]
- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [ ] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

I observe a lot of stale processes when I'm trying to train a model using `fit_generator()` with `use_multiprocessing=True`.

Key points:
- data has to be prepared with multiprocessing (`use_multiprocessing=True`);
- both train and validation sets have to be `Sequences`;
- main process has to be heavily loaded (the model has to be complex enough);
- I could not reproduce this behavior with `Dense` layer, but I have not spent much time on this.

I expect to see 5 python processes (1 main process and 4 worker processes) during the whole training. But as training proceeds, there are more and more processes. 5 of them are active (use CPU) while other are idle (stale). It seems to me that these are worker processes that for some reason failed to join the main process.

Usually the whole thing looks like this:
- 1 process at startup
- 5 processes during the first epoch
- on every next epoch the number of processes can increase by 1-4 processes

While legacy processes do not use the CPU, they use RAM, which leads to RAM shortage and eventually training failure.

Code to reproduce this issue:

    import os
    os.environ['TF_CPP_MIN_LOG_LEVEL'] = "2"
    
    import numpy as np
    from keras.utils import Sequence
    from keras.models import Model
    from keras.layers import Input, RepeatVector, GRU
    
    BATCH_SIZE = 64
    DIM_A = 50
    DIM_B = 26
    
    class MySequence(Sequence):
        def __len__(self):
            return 100
    
        def __getitem__(self, idx):
            sample_input = np.zeros((BATCH_SIZE, DIM_A, DIM_B), dtype=np.float32)
            sample_output = np.zeros((BATCH_SIZE, 1), dtype=np.float32)
            return sample_input, sample_output
    
    if __name__ == '__main__':
        inputs = Input(shape=(DIM_A, DIM_B))
        encoded = GRU(1, return_sequences=False)(inputs)
    
        model = Model(inputs, encoded)
        model.compile(optimizer='adam', loss='mean_squared_error')
    
        train_sequence = MySequence()
        validation_sequence = MySequence()
        model.fit_generator(
            train_sequence,
            len(train_sequence),
            epochs=100,
            verbose=1,
            validation_data=validation_sequence,
            validation_steps=len(validation_sequence),
            workers=4,
            use_multiprocessing=True)

System configuration:

- Ubuntu 16.04
- Keras (2.1.1) installed from git yesterday
- tensorflow-gpu (1.4.0)
- GeForce GTX 1080 GPU
- cuda-8-0 (8.0.61-1)
- libcudnn6 (6.0.21-1+cuda8.0)
- python3.5 (3.5.2-2ubuntu0~16.04.3)

**Comment by Dref360** [comment]
By stale process, you mean more process than you would imagine and they are not cleared or they are just not doing anything ie. no CPU usage?

**Comment by vsenko** [comment]
My fault, issue description is a bit obscure, I'll update it.

I expect to see 5 python processes (1 main process and 4 worker processes) during the whole training. But as training proceeds, there are more and more processes. 5 of them are active (use CPU) while other are idle (stale). It seems to me that these are worker processes that for some reason failed to join the main process.

Usually the whole thing looks like this:
- 1 process at startup
- 5 processes during the first epoch
- on every next epoch the number of processes can increase by 1-4 processes

While legacy processes do not use the CPU, they use RAM, which leads to RAM shortage and eventually training failure.

**Comment by tRosenflanz** [comment]
I am experiencing a similar issue. Might be not related but when trying to interrupt the training with multiprocessing enabled it frequently doesn't work and additional processes keep getting spawned replacing the ones you have just interrupted

**Comment by tRosenflanz** [comment]
I am also observing slow down of the training over several epochs and RAM usage getting progressively larger

**Comment by Dref360** [comment]
Right now, there is 2 possibilities.

1. The pool is not working properly. We should then use a context [first answer](https://stackoverflow.com/questions/18414020/memory-usage-keep-growing-with-pythons-multiprocessing-pool)

2. Reducing the queue size may help? Be aware that each element of the queue size is a worker working inside a process. Could yall try to pull master and try to reduce the queue_size to 5? I'm stiill trying to reproduce it.

**Comment by tRosenflanz** [comment]
@Dref360 I will try with the reduced queue but since my generator is very uneven (some batches take 
 much longer than others) I am working with larger queue to speed up the training. Permanently decreasing the queue size doesn't seem to be an optimal solution

**Comment by Dref360** [comment]
Yes I understand, it's just to point me in the right direction. :) Thx for your help.

**Comment by Dref360** [comment]
If anyone is able to test this it would be awesome, I added a closing context so at the end of each epoch, the pool is guaranteed (I think) to be closed.

https://github.com/Dref360/keras/commit/d2e51a84b74be6a5df74fcbf240557bcd853832c

**Comment by vsenko** [comment]
I'll have an opportunity to run tests on monday.

3 дек. 2017 г. 3:02 пользователь "Frédéric Branchaud-Charron" <
notifications@github.com> написал:

If anyone is able to test this it would be awesome, I added a closing
context so at the end of each epoch, the pool is guaranteed (I think) to be
closed.

Dref360/keras@d2e51a8
<https://github.com/Dref360/keras/commit/d2e51a84b74be6a5df74fcbf240557bcd853832c>

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/fchollet/keras/issues/8622#issuecomment-348719825>, or mute
the thread

**Comment by Dref360** [comment]
This has been merged to fix something else, so just pull master and it should work (I hope )

**Comment by tRosenflanz** [comment]
@Dref360 I tested with smaller number of steps per epoch and the issue seems to be resolved. Now testing with the full length which might take a bit of time but I don't think it will make a difference.

**Comment by vsenko** [comment]
I have also rerun small tests and everything looks fine! Currently I'm experimenting with other code, it'll take some time.

Some more good news: it seems like #8619 is also fixed with update!

**Comment by vsenko** [comment]
@Dref360, thank you! All my test ran OK, hence I'm closing this issue.
